chain-impl-mockchain/benches/vote/vote_cast.rs

```diff
@@ -0,0 +1,105 @@
+use super::VoteBenchmarkConfig;
+use chain_crypto::testing::TestCryptoRng;
+use chain_impl_mockchain::{
+    certificate::{VoteCast, VotePlan},
+    testing::VoteTestGen,
+    vote::Choice,
+};
+use criterion::{criterion_group, Criterion};
+use rand::{
+    distributions::{Bernoulli, Distribution},
+    SeedableRng,
+};
+
+fn vote_cast_benchmark<D: Distribution<u64>>(
+    benchmark_name: &str,
+    config: VoteBenchmarkConfig<D>,
+    c: &mut Criterion,
+) {
+    let mut rng = TestCryptoRng::seed_from_u64(0);
+    let (ledger, controller, _) = config.setup(&mut rng);
+
+    // cast votes
+    let vote_plan_def = controller.vote_plan(super::VOTE_PLAN).unwrap();
+    let vote_plan: VotePlan = vote_plan_def.into();
+
+    let should_vote_dist = Bernoulli::new(config.proposals_per_voter_ratio).unwrap();
+    let yes_vote_dist = Bernoulli::new(config.yes_votes_ratio).unwrap();
+    let vote_fragments = {
+        let mut res = Vec::new();
+        for (proposal_idx, proposal) in vote_plan.proposals().iter().enumerate() {
+            for private_voter in controller.wallets().iter().skip(1) {
+                let should_vote = should_vote_dist.sample(&mut rng);
+                if !should_vote {
+                    continue;
+                }
+
+                let choice = Choice::new(yes_vote_dist.sample(&mut rng) as u8);
+                let payload = VoteTestGen::private_vote_cast_payload_for(
+                    &vote_plan, proposal, choice, &mut rng,
+                );
+
+                res.push(controller.fragment_factory().vote_cast(
+                    private_voter,
+                    VoteCast::new(vote_plan.to_id(), proposal_idx as u8, payload),
+                ));
+            }
+        }
+        res
+    };
+
+    c.bench_function(&format!("vote_cast_{}", benchmark_name), |b| {
+        b.iter(|| {
+            let mut ledger = ledger.clone();
+            for vote in &vote_fragments {
```

Comment:
It looks like I had a wrong impression of what you were going to measure. For this kind of benchmark it is fine to include it into the existing benchmark function. It can be filtered out if needed.

Reply:
What did you mean? (maybe it's worthwhile to add that as well :))

Reply:
I mean that you generate exactly the same data we use for the vote tally benchmark, so you can just insert that benchmark in the sequence of benchmarking the vote tally. If someone does not want to run this benchmark, and only benchmark the tally steps, they can always do this: `cargo bench --manifest-path chain-impl-mockchain/Cargo.toml --features with-bench -- tally`

Reply:
Yes, that makes sense, but you hinted that there was miscommunication about the subject of the benchmark (i.e. you thought to benchmark another part) and maybe we could add that as well

Reply:
I thought you were going to benchmark one individual vote, not the whole fund load.

Reply:
This can be discussed indeed. I though to benchmark the whole fund because the number of previous votes affect the size of the hamt used to store them for example, so I suspect that adding the 1e6th vote could be slower. I may as well look into that to see if we can avoid to benchmark the whole fund

Reply:
> I may as well look into that to see if we can avoid to benchmark the whole fund

Benchmarking the whole fund is perfectly fine, because this

> the number of previous votes affect the size of the hamt used to store them

is a perfectly valid concern.

I meant that I misunderstood what you wanted to do and gave you the wrong advice about structuring your benchmarks.